pulsar-broker/pom.xml

```diff
@@ -274,6 +274,14 @@
       <version>${project.version}</version>
       <scope>test</scope>
     </dependency>
+    
+    <dependency>
```

Comment:
it looks a bit strange to put connector in broker for testing purpose.

if you need some test cases from pulsar-broker, it would be much better that pulsar-broker generates a test-jar. so pulsar-io-kinesis can use that test-jar to write its test.

Reply:
right now, it seems we don't have unit test which can test pulsar-broker and pulsar-function. having this e2e test help to find out many unknown issues and also help testing while development. we also have admin-rest-api for functions at pulsar-broker module for which we don't have test. so, this will be used to add e2e test which can cover all those code paths.

> if you need some test cases from pulsar-broker, it would be much better that pulsar-broker generates a test-jar. 

umm.. not exactly. as I explained above, the idea is to start mock broker and fun-worker and test kinesis-sink there. so, we might not be able to write it at kinesis-sink module.
Hope that make sense?

Reply:
> having this e2e test help to find out many unknown issues and also help testing while development.

yes, we need to have this e2e test. actually internally we have developed the e2e tests using testcontainers (driven by @aahmed-se ), we will start pushing out the integration tests to apache soon.


I think functions/proxy/websocket are kind of different from a specific connector implementation. functions/proxy/websocket are generic components for pulsar, which they are okay to be included in broker since they provides different functionalities of pulsar. however kinesis connector is an implementation of the pulsar I/O framework, which doesn't really belong to broker component. 

image you are a connector developer, you develop a connector system `xyz`. are you expecting to put this connector into broker module and test it? or does it sound better if the connector framework provides a test base, that you can include in your connector code and test the connector using the test base.

hope this clarify my points here.

Reply:
> image you are a connector developer, you develop a connector system xyz. are you expecting to put this connector into broker module and test it?

Make sense. Actually, I was trying to add e2e test for broker + function which we don't have and I picked up kinesis-sink in test-case as an example. may be I can replace kinesis-sink with pulsar-sink to keep it more generic e2e test. will that be fine?

Reply:
using pulsar-sink is fine to me.

Reply:
@sijie I have added `PulsarSinkTets` . However, I feel keeping Kinesis-Connector e2e test is not harmful as it's just a test and dependency-scope is also test. I think it is helping to add more test code-coverage in pulsar-function where existing testcases are not touching some of the code paths. We can definitely remove it in future and only keep PulsarSinkTest if it is causing any issue for dependency management but until we can keep it as it will be helpful for future development testing.

Reply:
I have to agree with @sijie here. This one just breaks the broker abstraction.

Reply:
> This one just breaks the broker abstraction.

Umm..But function also depends on broker for the execution. so, e2e test will require broker init it. and adding test will not impact any code level abstraction. ?

Reply:
@rdhabalia do we need to put the tests in the pulsar-broker module?  Can we not just put in the module for pulsar-io/kinesis ?  We can add a test scope dependency to the broker there?

Reply:
actually, in order to add it individual source/sink, we need to adapt broker and functions mock-service tests into those individual modules. Pulsar-broker starts services for broker, websocket, proxy and functions as well so, we have kept all e2e tests for services into broker-modules and no need to copy mocks at individual service level.